File: codes/0670-Maximum-Swap/Java/main3.java
class Solution {
    public int maximumSwap(int num) {
        char[] arr = String.valueOf(num).toCharArray();
        for(int i = 0; i < arr.length; ++i) {
            int maxIndex = i;
            for(int j = arr.length-1; j > i; --j) {
                if(arr[j] > arr[maxIndex]) {
                    maxIndex = j;
                }
            }
            if(maxIndex > i) {
                char t = arr[i];
                arr[i] = arr[maxIndex];
                arr[maxIndex] = t;
                return Integer.valueOf(new String(arr));
            }
        }
        return num;
    }
}

File: codes/0670-Maximum-Swap/cpp/main3.cpp
class Solution
{
public:
    int maximumSwap(int num)
    {
        string str = to_string(num);
        vector<int> buckets(10, 0);
        for (int i = 0; i < str.size(); ++i)
        {
            buckets[str[i] - '0'] = i;
        }
        for (int i = 0; i < str.size(); ++i)
        {
            for (int k = 9; k > str[i] - '0'; --k)
            {
                if (buckets[k] <= i)
                    continue;
                swap(str[i], str[buckets[k]]);
                return stoi(str);
            }
        }
        return num;
    }
};